### PR TITLE
UrlMatchPattern: Add support for port match

### PR DESCRIPTION
**Pull request by codenirvana**
- Match HOST and PORT separately
- Implicitly match default PORTs for a given PROTOCOL
- Fixed a bug where empty HOST are valid

### CODE REVIEW

File: test/unit/certificate.test.js [hunk 1]
```diff
@@ -53,6 +53,38 @@ describe('Certificate', function () {
             expect(certificate.canApplyTo('http://www.google.com')).to.be.false;
             expect(certificate.canApplyTo('https://www.google.com')).to.be.false;
         });
+
+        it('should match given port correctly', function () {
+            var certificate = new Certificate({ matches: ['https://example.com:8080/*'] });
+            expect(certificate.canApplyTo('https://example.com:443')).to.be.false;
+            expect(certificate.canApplyTo('https://example.com:8080')).to.be.true;
+
+            certificate = new Certificate({ matches: ['https://example.com:*/*'] });
```
Comment (kunagpal) [inline test/unit/certificate.test.js]: This should be its own test.

File: lib/url-pattern/url-match-pattern.js [hunk 2]
```diff
@@ -187,6 +200,38 @@ _.assign(UrlMatchPattern.prototype, /** @lends UrlMatchPattern.prototype */ {
...
+        // return true if both given port and match pattern are absent
+        if (typeof port === UNDEFINED && typeof portRegex === UNDEFINED) {
+            return true;
+        }
+
+        // convert integer port to string
+        (port && typeof port !== STRING) && (port = String(port));
+
+        // assign default port or portRegex
+        port = port || defaultPort;
```
Comment (kunagpal) [inline lib/url-pattern/url-match-pattern.js]:
```js
!port && (port = defaultPort);
```

Same goes below.

File: test/unit/certificate.test.js [hunk 3]
```diff
@@ -53,6 +53,38 @@ describe('Certificate', function () {
             expect(certificate.canApplyTo('http://www.google.com')).to.be.false;
             expect(certificate.canApplyTo('https://www.google.com')).to.be.false;
         });
+
+        it('should match given port correctly', function () {
+            var certificate = new Certificate({ matches: ['https://example.com:8080/*'] });
```
Comment (kunagpal) [inline test/unit/certificate.test.js]: Newline after every `var` block.

File: test/unit/url-match-pattern.test.js [hunk 4]
```diff
@@ -309,6 +330,49 @@ describe('UrlMatchPattern', function () {
...
+            var matchPattern = new UrlMatchPattern('https://*:1234/*');
+            expect(matchPattern.testPort(1234)).to.be.true;
+            expect(matchPattern.testPort('1234')).to.be.true;
+            expect(matchPattern.testPort(4321)).to.be.false;
+        });
+
+        it('should implicitly match with default http port if protocol is defined', function () {
+            var matchPattern = new UrlMatchPattern('http://*/*');
+            expect(matchPattern.testPort(undefined, 'http')).to.be.true;
+            expect(matchPattern.testPort('80', 'http')).to.be.true;
```
Comment (kunagpal) [inline test/unit/url-match-pattern.test.js]: Numerical `80` as well.

File: lib/url-pattern/url-match-pattern.js [hunk 5]
```diff
@@ -11,15 +14,24 @@ var _ = require('../util').lodash,
     ALLOWED_PROTOCOLS = ['http', 'https', 'file', 'ftp'],
     ALLOWED_PROTOCOLS_REGEX = ALLOWED_PROTOCOLS.join('|'),
 
+    DEFAULT_PROTOCOL_PORT = {
```
[the remark this answers is not in the document]
Comment (codenirvana) [reply lib/url-pattern/url-match-pattern.js]: Added as todo, we will initialize protocols while initializing the `UrlMatchPattern`

**Pull request merged**